agent/agent.go

```diff
@@ -0,0 +1,825 @@
+		}
+		lastChunkIdx = chunkIdx
+
+		numChunks++
+		bytes := request.GetData().GetBytes()
+		checksum = crc32.Update(checksum, crc32.IEEETable, bytes)
+
+		numWritten, err := fileHandle.write(bytes)
+		if err != nil {
+			return err
+		}
+
+		if numWritten != len(bytes) {
+			return fmt.Errorf("unable to write bytes, expected: %d, observed: %d", len(bytes), numWritten)
+		}
+
+		if streamErr == io.EOF {
+			return doneFn(fileHandle, fileType, checksum, numChunks)
+		}
+	}
+}
+
+type opAgentHeartBeater struct {
+	sync.RWMutex
+	iopts             instrument.Options
+	agent             *opAgent
+	running           bool
+	wg                sync.WaitGroup
+	timeoutCh         chan struct{}
+	msgChan           chan heartbeatMsg
+	conn              *grpc.ClientConn
+	client            hb.HeartbeaterClient
+	operatorUUID      string
+	heartbeatEndpoint string
+	lastHeartbeatTs   time.Time
+	nowFn             xclock.NowFn
+}
+
+func newHeartbeater(
+	agent *opAgent,
+	timeoutCh chan struct{},
+	opUUID string,
+	hbEndpoint string,
+	iopts instrument.Options,
+) (*opAgentHeartBeater, error) {
+	conn, err := grpc.Dial(hbEndpoint, grpc.WithTimeout(defaultHeartbeatConnTimeout), grpc.WithInsecure())
+	if err != nil {
+		return nil, err
+	}
+	client := hb.NewHeartbeaterClient(conn)
+	return &opAgentHeartBeater{
+		iopts:             iopts,
+		agent:             agent,
+		timeoutCh:         timeoutCh,
+		msgChan:           make(chan heartbeatMsg),
+		conn:              conn,
+		client:            client,
+		operatorUUID:      opUUID,
+		heartbeatEndpoint: hbEndpoint,
+		nowFn:             agent.opts.NowFn(),
+	}, nil
+}
+
+func (h *opAgentHeartBeater) isRunning() bool {
+	h.RLock()
+	defer h.RUnlock()
+	return h.running
+}
+
+func (h *opAgentHeartBeater) defaultHeartbeat() hb.HeartbeatRequest {
+	h.RLock()
+	req := hb.HeartbeatRequest{
+		OperatorUuid:   h.operatorUUID,
+		ProcessRunning: h.agent.Running(),
+	}
+	h.RUnlock()
+	return req
+}
+
+func (h *opAgentHeartBeater) start(d time.Duration) {
+	h.Lock()
+	h.running = true
+	h.Unlock()
+	h.wg.Add(1)
+	go h.heartbeatLoop(d)
+}
+
+func (h *opAgentHeartBeater) sendHealthyHeartbeat() {
+	beat := h.defaultHeartbeat()
+	beat.Code = hb.HeartbeatCode_HEALTHY
+	h.sendHeartbeat(&beat)
+}
+
+func (h *opAgentHeartBeater) heartbeatLoop(d time.Duration) {
+	defer func() {
+		h.Lock()
+		h.running = false
+		h.Unlock()
+		h.wg.Done()
+	}()
+
+	// explicitly send first heartbeat as soon as we start
+	h.sendHealthyHeartbeat()
+
+	for {
+		select {
+		case msg := <-h.msgChan:
+			if msg.stop {
+				return
+			}
+			beat := h.defaultHeartbeat()
+			msg.toRPCType(&beat)
+			h.sendHeartbeat(&beat)
+
+		default:
+			h.sendHealthyHeartbeat()
+			// NB(prateek): we use a sleep instead of a ticker because the latter
+			// only guarantees a lower bound on the frequency of ticks, it does not
+			// guarantee any upper bound. This makes it impossible to reliably detect
+			// heartbeating timeouts.
```

Comment:
I don't believe `time.Sleep` guarentees an upper bound either.  It is up to the Go runtime goroutine scheduler between how much lag time there is after `time.Sleep(...)` call actually wakes up after the duration passed in.  

Given that these will be very similar, it seems better to use a ticker so that it can be stopped mid-tick no?  Right now sending a stop message needs to wait a whole tick if it just started it's `time.Sleep(...)` i.e.
```go
tick := time.NewTicker(d)
defer tick.Stop()
for {
  select { 
  case msg := <-msg.Chan:
    //handle msg
  case <-tick.C:
    // 
  }
}
```